src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+        t = np.arange(self.device.adc_sample_size) / adc_sampling_rate
+        results = {}
+
+        for readout_pulse in sequence:
+            qubit = qubits[readout_pulse.qubit]
+            _, adc = qubit.readout.ports
+
+            raw_signal = adc_raw_data[adc]
+            sin = np.sin(2 * np.pi * readout_pulse.frequency * t)
+            cos = np.sin(2 * np.pi * readout_pulse.frequency * t)
+
+            I = np.dot(raw_signal, cos)
+            Q = np.dot(raw_signal, sin)
+            results[readout_pulse.qubit] = IntegratedResults(I + 1j * Q)
+
+            if options.averaging_mode is not AveragingMode.SINGLESHOT:
+                results[readout_pulse.qubit] = results[readout_pulse.serial] = results[
+                    readout_pulse.qubit
+                ].average
+            else:
+                results[readout_pulse.serial] = results[readout_pulse.qubit]
+
+        return results
+
+    def sweep(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequence: PulseSequence,
+        options: ExecutionParameters,
+        *sweeper: Sweeper,
+    ):
+        # Record pulse values before sweeper modification
+        bsv = []
+        for sweep in sweeper:
+            param_name = sweep.parameter.name.lower()
+            base_sweeper_values = [getattr(pulse, param_name) for pulse in sweep.pulses]
+            bsv.append(base_sweeper_values)
+
+        res = self._sweep(qubits, couplers, sequence, options, *sweeper)
+
+        # Reset pulse values back to original values
+        for sweep, base_sweeper_values in zip(sweeper, bsv):
+            param_name = sweep.parameter.name.lower()
+            for pulse, value in zip(sweep.pulses, base_sweeper_values):
+                setattr(pulse, param_name, value)
+
+                if pulse.type is PulseType.READOUT:
+                    res[pulse.serial] = res[pulse.qubit]
+
+        return res
+
+    def _sweep(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequence: PulseSequence,
+        options: ExecutionParameters,
+        *sweeper: Sweeper,
+    ):
+        """Recursive python-based sweeper functionaltiy for the IcarusQ
+        RFSoC."""
+        if len(sweeper) == 0:
+            return self.play(qubits, couplers, sequence, options)
+
+        sweep = sweeper[0]
+        param = sweep.parameter
+        param_name = param.name.lower()
+
+        if param not in self.available_sweep_parameters:
+            raise NotImplementedError(
+                "Sweep parameter requested not available", param_name
+            )
+
+        base_sweeper_values = [getattr(pulse, param_name) for pulse in sweep.pulses]
+        sweeper_value_setter_fn = _sweeper_value_setter.get(sweep.type)
+        ret = {}
+
+        for value in sweep.values:
+            for idx, pulse in enumerate(sweep.pulses):
+                base = base_sweeper_values[idx]
+                sweeper_value_setter_fn(pulse, param_name, value, base)
```

Comment:
Actually, here you could simplify a bit, since part of the implementation is the same for all the types:
```suggestion
                
```
```suggestion
        sweeper_op = _sweeper_operation.get(sweep.type)
        ret = {}

        for value in sweep.values:
            for idx, pulse in enumerate(sweep.pulses):
                base = base_sweeper_values[idx]
                setattr(pulse, param_name, sweeper_op(value, base))
```

Reply:
In which case, you could use [`operator.add`](https://docs.python.org/3/library/operator.html#operator.add) and [`operator.mul`](https://docs.python.org/3/library/operator.html#operator.mul), without defining `lambda`s

Reply:
Resolved in b87ea2c